modules/javafx.base/src/test/java/test/javafx/beans/value/ObservableValueFluentBindingsTest.java

```diff
@@ -878,6 +884,284 @@ void shouldNoLongerBeStronglyReferenced() {
+                        ReferenceAsserts.testIfNotStronglyReferenced(observableValue, () -> {
+                            observableValue = null;
+                            condition = null;
+                        });
+                    }
+                }
+            }
+
+            @Nested
+            class WhenObservedForChanges {
+                {
+                    startObservingChanges(observableValue);
+                }
+
+                @Test
+                void shouldReceiveCurrentPropertyValues() {
+                    assertNothingIsObserved();
+
+                    property.set("Right");
+
+                    assertObserved("Right");
+                }
+
+                @Test
+                void shouldOnlyReceiveCurrentPropertyValuesWhileConditionIsTrue() {
+                    assertNothingIsObserved();
+
+                    property.set("Right");
+
+                    assertObserved("Right");
+
+                    condition.set(false);
+
+                    assertNothingIsObserved();
+
+                    property.set("Left");
+
+                    assertNothingIsObserved();
+
+                    property.set("Middle");
+
+                    assertNothingIsObserved();
+
+                    condition.setValue(null);  // null is false as well, should not change result
+
+                    assertNothingIsObserved();
+
+                    condition.set(true);
+
+                    assertObserved("Middle");
+                }
+
+                @Test
+                void shouldBeStronglyReferenced() {
+                    ReferenceAsserts.testIfStronglyReferenced(observableValue, () -> {
+                        observableValue = null;
+                        condition = null;
+                    });
+                }
+
+                @Test
+                void shouldNotBeStronglyReferencedWhenConditionIsFalse() {
+                    condition.set(false);
+
+                    ReferenceAsserts.testIfNotStronglyReferenced(observableValue, () -> {
+                        observableValue = null;
+                        condition = null;
+                    });
+                }
+
+                @Nested
+                class AndWhenUnobserved {
+                    {
+                        stopObservingChanges(observableValue);
+                    }
+
+                    @Test
+                    void shouldNoLongerBeCalled() {
+                        assertNothingIsObserved();
+
+                        property.set("Right");
+
+                        assertNothingIsObserved();
+                    }
+
+                    @Test
+                    void shouldNoLongerBeStronglyReferenced() {
+                        ReferenceAsserts.testIfNotStronglyReferenced(observableValue, () -> {
+                            observableValue = null;
+                            condition = null;
+                        });
+                    }
+                }
+            }
+        }
```

Comment:
Is it possible to add the scenario that simulates configurations you so beautifully illustrated in the comments in this PR?  For example, simulate the removal of the "Skin" (without involving a real Skin implementation, just re-create the classes and properties in the same configuration)?

Reply:
You mean something like calling `setSkin` on a `Button` in a loop with a new skin that uses this form of listener management?  And seeing if the old skin instances don't stick around?

I suppose I could make a copy of one of the smaller skins, tweak it to use `when` and then testing this. Is that what you'd be looking for?

Reply:
never mind, please disregard.